cdc/capture/capture.go

```
@@ -0,0 +1,349 @@
+	return &Capture{
+		pdClient:   pdClient,
+		kvStorage:  kvStorage,
+		etcdClient: etcdClient,
+		cancel:     func() {},
+
+		newProcessorManager: processor.NewManager,
+		newOwner:            owner.NewOwner,
+	}
+}
+
+func (c *Capture) reset() error {
+	c.captureMu.Lock()
+	defer c.captureMu.Unlock()
+	conf := config.GetGlobalServerConfig()
+	c.info = &model.CaptureInfo{
+		ID:            uuid.New().String(),
+		AdvertiseAddr: conf.AdvertiseAddr,
+		Version:       version.ReleaseVersion,
+	}
+	c.processorManager = c.newProcessorManager()
+	if c.session != nil {
+		c.session.Close() //nolint:errcheck
+	}
+	sess, err := concurrency.NewSession(c.etcdClient.Client.Unwrap(),
+		concurrency.WithTTL(conf.CaptureSessionTTL))
+	if err != nil {
+		return errors.Annotate(cerror.WrapError(cerror.ErrNewCaptureFailed, err), "create capture session")
+	}
+	c.session = sess
+	c.election = concurrency.NewElection(sess, kv.CaptureOwnerKey)
+	log.Info("init capture", zap.String("capture-id", c.info.ID), zap.String("capture-addr", c.info.AdvertiseAddr))
+	return nil
+}
+
+// Run runs the capture
+func (c *Capture) Run(ctx context.Context) error {
+	defer log.Info("the capture routine has exited")
+	for {
+		select {
+		case <-ctx.Done():
+			return nil
+		default:
+		}
+		err := c.reset()
+		if err != nil {
+			return errors.Trace(err)
+		}
+		ctx, cancel := context.WithCancel(ctx)
+		c.cancel = cancel
+		err = c.run(ctx)
+		// if capture suicided, reset the capture and run again.
+		// if the canceled error throw, there are two possible scenarios:
+		//   1. the internal context canceled, it means some error happened in the internal, and the routine is exited, we should restart the capture
+		//   2. the parent context canceled, it means that the caller of the capture hope the capture to exit, and this loop will return in the above `select` block
+		if cerror.ErrCaptureSuicide.Equal(err) || context.Canceled == errors.Cause(err) {
```

if err is `context.Canceled`, the following `capture recovered` log is not needed?

---

if the internal context canceled, the capture recovered log is needed;
if the parent context canceled, the capture recovered log is not needed;
What suggestions do you have to refine it?

---

IMHO the internal cancel should return the real error instead of context.Canceled